src/rabbitmq.js

```diff
@@ -0,0 +1,354 @@
+  /**
+   * Unsubscribe and stop consuming from all queues.
+   *
+   * @return {Promise} Promise resolved when all queues canceled.
+   */
+  unsubscribe (): Promise {
+    const consuming = this.consuming
+    return Promise.map(consuming.keySeq(), (queue) => {
+      const consumerTag = consuming.get(queue)
+      return Promise.resolve(this.channel.cancel(consumerTag))
+        .then(() => {
+          this.consuming = this.consuming.delete(queue)
+        })
+    })
+  }
+
+  /**
+   * Disconnect from RabbitMQ.
+   *
+   * @return {Promise} Promise resolved when disconnected from RabbitMQ.
+   */
+  disconnect (): Promise {
+    if (!this._isPartlyConnected()) {
+      return Promise.reject(new Error('not connected. cannot disconnect.'))
+    }
+    return Promise.resolve(this.connection.close())
+  }
+
+  // Private Methods
+
+  /**
+   * Error handler for the RabbitMQ connection.
+   *
+   * @private
+   * @throws Error
+   * @param {object} err Error object from event.
+   */
+  _connectionErrorHandler (err: Error) {
+    this.log.error({ err: err }, 'connection has caused an error')
+    throw err
+  }
+
+  /**
+   * Error handler for the RabbitMQ channel.
+   * @private
+   * @throws Error
+   * @param {object} err Error object from event.
+   */
+  _channelErrorHandler (err: Error) {
+    this.log.error({ err: err }, 'channel has caused an error')
+    throw err
+  }
+
+  /**
+   * Check to see if model is connected.
+   *
+   * @return {Boolean} True if model is connected and channel is established.
+   */
+  _isConnected (): boolean {
+    return !!(this._isPartlyConnected() && this.channel)
+  }
+
+  /**
+   * Check to see if model is _partially_ connected. This means that the
+   * connection was established, but the channel was not.
+   *
+   * @return {Boolean} True if connection is established.
+   */
+  _isPartlyConnected (): boolean {
+    return !!(this.connection)
+  }
+
+  /**
+   * Helper function to consolidate logic for subscribing to queues. Stores
+   * information about what is subscribed and is responsible for asserting
+   * exchanges and queues into existance.
+   *
+   * @private
+   * @param {Object} opts Object describing the exchange connection.
+   * @param {String} opts.exchange Name of exchange.
+   * @param {String} opts.handler Handler of jobs.
+   * @param {String} opts.type Type of exchange: 'fanout' or 'topic'.
+   * @param {String} [opts.routingKey] Routing key for a topic exchange.
+   * @return {Promise} Promise resolved when subcribed to exchange.
+   */
+  _subscribeToExchange (opts: SubscribeObject): Promise {
+    const log = this.log.child({
+      method: '_subscribeToExchange',
+      opts: opts
+    })
+    log.info('subscribing to exchange')
+    if (!this._isConnected()) {
+      return Promise.reject(new Error('must .connect() before subscribing'))
+    }
+    if (opts.type === 'topic' && !opts.routingKey) {
+      return Promise.reject(new Error('routingKey required for topic exchange'))
```

Comment:
is routing key actually required? what if I want all topics ? am I forced to pass '*' ?